### DISCUSSION TITLE
Reading an image before and after an external modification

### DISCUSSION THREAD

**Discussion by MathemanFlo** [libvips/libvips#2760] [Help wanted]
Hello John,
I have yet another small issue with libvips.
In short words I load an image twice but inbetween the image gets modified by another program. Example:
```c++
#include <iostream>
#include <vips/vips.h>
#include <vips/vips8>

using namespace std;
using namespace vips;

int main(int argc, char** argv)
{
    VIPS_INIT(argv[0]);
    VImage img = VImage::new_from_file("image.tif");
    img.write_to_file("aa.tif");

    cout << "Modify image with an external program" << endl;
    getchar();

    VImage img2 = VImage::new_from_file("image.tif");
    img2.write_to_file("aa.tif");
    return 0;
}
```
My problem is now that libvips reads two times the old, unmodified image, because the file is cached and not reread. How can I force libvips to reread the image? I tried clearing the cache via `vips_cache_drop_all()` but calling the functions lets the second `new_from_file` crash.
Thank you for your time and help in advance.

**Comment by jcupitt** [comment]
Hi @MathemanFlo, I think you need `invalidate`:

https://www.libvips.org/API/current/VipsImage.html#vips-image-invalidate-all

Use that on an image and that image and all downstream images will be kicked out of cache.

**Reply by MathemanFlo** [reply]
Thank you very much. That works for me, thank you for the explanation, you really saved me a lot of time. While reading the documentation I read: 

> You should call this function after destructively modifying an image with something like [vips_draw_circle()](https://www.libvips.org/API/current/libvips-draw.html#vips-draw-circle).
It's nowhere mentioned at https://www.libvips.org/API/current/libvips-draw.html

Maybe it is also worth mentioning the invalidate mechanism at a proper place. I'm not sure where though...

**Reply by jcupitt** [reply]
The draw docs say to make a copy of the image first, so there's no need to invalidate in this case.

Yes, I'm not sure where it should be mentioned.

**Reply by MathemanFlo** [reply]
Sorry for the noise. I was 100% sure the issue is fixed by calling `vips_image_invalidate_all` but actually it isn't. I put `vips_image_invalidate_all(img.get_image());` and `vips_image_invalidate_all(img2.get_image());` at any possible place but the issue persists.

**Comment by MathemanFlo** [comment]
Hi @jcupitt 
I did quite a few further experiments on this and I still face some problems.
```c++
  VImage img = VImage::new_from_file("image.tif");
  img.write_to_file("first.tif");
  cout << "Modify image with an external program" << endl;
  getchar();
  vips_cache_print();
  VImage img2 = VImage::new_from_file("image.tif");
  vips_image_invalidate_all(img2.get_image());
  vips_cache_print(); // still prints the very same
  img2.write_to_file("second.tif");
```
second.tif is the same image as first.tif regardless any modifications in between.

My second attempt was to clear the cache. Calling `vips_cache_drop_all();` after the `getchar()` makes the program crash on `new_from_file` because `vips_cache_operation_buildp` in `VImage::call_option_string` does not succeed.
Thus I cleared the cache after the `getchar()`  by 
```c++
int maxCache = vips_cache_get_max();
vips_cache_set_max(0);
vips_cache_set_max(maxCache);
```
before loading the image the second time. This works indeed and would be a work around but I do not understand why my first attempt is unsuccessful. I also tried to invalidate `img` (which does not make too much sense according to my understanding) but with no success...
